### PR TITLE
Allow users to specify metric success function with ArmeriaAutoConfiguration

### PR DESCRIPTION
**Pull request by nirvanarsc**
This PR relates to https://github.com/line/armeria/pull/3410

### Motivation
- Currently the custom success function is not compatible with ArmeriaAutoConfiguration, so allow users to configure it by defining a `@Bean` returning a MetricCollectingConfigurator.

### Modifications
- Added a `MetricCollectingConfigurator`
- Cleanup `HelloConfiguration.java`, remove the `MetricCollectingService` decorator, since its incompatible with ArmeriaAutoConfiguration, and we also don't have meter registries in the minimal example project.
- Added a unit test in `ArmeriaAutoConfigurationTest`

### CODE REVIEW

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/MetricCollectingConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Interface used to configure a {@code MetricCollectingService} on the default Armeria server.
+ */
+@FunctionalInterface
+public interface MetricCollectingConfigurator extends Ordered {
+
+    /**
+     * Configures the {@link MetricCollectingService} using the specified {@link AbstractMetricCollectingBuilder}.
+     */
+    void configure(AbstractMetricCollectingBuilder metricCollectingServiceBuilder);
```
Comment (nirvanarsc) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/MetricCollectingConfigurator.java:36]: Although we only configure `MetricCollectingServiceBuilder` in `ArmeriaConfigurationUtil.java` I'm using the super class here. `MetricCollectingClientBuilder` and `MetricCollectingRpcClientBuilder` can be configured separately

File: spring/boot2-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaAutoConfigurationTest.java [hunk 2]
```diff
@@ -349,6 +365,26 @@ public void testMetrics() {
...
+        assertThat(res.contentUtf8()).isEqualTo("error");
+
+        final String metricReport = WebClient.of(newUrl("http"))
+                                             .get("/internal/metrics")
+                                             .aggregate().join()
+                                             .contentUtf8();
+        assertThat(metricReport).contains(
+                "http_status=\"404\",method=\"error\",result=\"success\",service=\"annotatedService\",} 1.0");
+        assertThat(metricReport).contains(
+                "http_status=\"404\",method=\"error\",result=\"failure\",service=\"annotatedService\",} 0.0");
```
Comment (ikhoon) [inline spring/boot2-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaAutoConfigurationTest.java:385]:
```suggestion
        await().untilAsserted(() -> {
            assertThat(metricReport).contains(
                    "http_status=\"404\",method=\"error\",result=\"success\",service=\"annotatedService\",} 1.0");
            assertThat(metricReport).contains(
                    "http_status=\"404\",method=\"error\",result=\"failure\",service=\"annotatedService\",} 0.0");
        });
```

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/MetricCollectingConfigurator.java (hunk 1 above)
Comment (ikhoon) [reply spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/MetricCollectingConfigurator.java:37]: > MetricCollectingClientBuilder and MetricCollectingRpcClientBuilder can be configured separately

Do you want to configure the client builders with `@Qualifier` annotation? I prefer to use `MetricCollectingServiceBuilder` for clarification. I think this should be only used for `MetricCollectingService`.

Comment (nirvanarsc) [reply spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/MetricCollectingConfigurator.java]: Hmm yes you are right, I want to let users use the builder classes and not the `@Qualifier` annotation. So I changed the type here 👍

**Review by trustin — APPROVED** [review]
Looks like a nice addition to me. :+1:

**Review by minwoox — APPROVED** [review]
Awesome! Thanks a lot. 😄

**Review by ikhoon — APPROVED** [review]
Thanks! @nirvanarsc

**Pull request merged**